File: frontend/src/stores/bookings.ts
import { ref } from 'vue'
import { defineStore } from 'pinia'
import { api } from '@/lib/api'
import { mockBookings, mockIncomingBookings } from '@/mocks/bookings'

export type BookingStatus = 'pending' | 'accepted' | 'declined' | 'completed'
export type PaymentMethod = 'coins' | 'card'

export interface BookingAddon {
  id: string
  label: string
  priceCoins: number
}

export interface Booking {
  id: string
  /** Human-facing order number shown on Checkout / Order Confirmation, e.g. "SQ-84213". */
  orderNumber: string
  playerId: string
  /** Denormalized display fields the backend joins in (`GET /bookings/...`) so list/detail
   * views don't need a separate per-row player fetch. Absent on the static mock fixtures, which
   * views fall back to `mocks/players.ts` lookups for. */
  playerDisplayName?: string
  playerAvatarUrl?: string | null
  serviceId: string
  serviceName?: string
  userId: string
  buyerDisplayName?: string
  status: BookingStatus
  serviceTypeLabel: string
  /** Coin price per unit (per game/hour/session, matching `priceUnit`). */
  priceCoins: number
  priceUnit: string
  quantity: number
  addons: BookingAddon[]
  promoLabel: string | null
  subtotalCoins: number
  addonsCoins: number
  discountCoins: number
  totalCoins: number
  paymentMethod: PaymentMethod
  /** `null` means "start now"; otherwise an ISO timestamp for a scheduled start. */
  scheduledFor: string | null
  createdAt: string
  /** Whether a review already exists for this booking (`POST /reviews` is one-per-booking) -
   * gates My Bookings' "Leave review" action. Absent on the static mock fixtures, which fall
   * back to `false` (never reviewed). */
  hasReview?: boolean
}

export interface ReviewPayload {
  rating: number
  text: string
}

/** A "Book a session" selection that hasn't been submitted yet - kept client-side only until
 * Checkout's "Place order" posts it, so an abandoned checkout never creates a real `pending`
 * row (see CHECKPOINT.md's Booking flow note). */
export interface BookingDraft {
  playerId: string
  /** Carried along so Checkout/Order Confirmation can render the Pal's name/service title
   * without a `mocks/players.ts` lookup, which only knows the seed Pals (`p1`..`p8`), not real
   * ones fetched via `GET /players` (3.2). */
  playerDisplayName: string
  serviceId: string
  serviceName: string
  serviceTypeLabel: string
  priceCoins: number
  priceUnit: string
  quantity: number
  addons: BookingAddon[]
  promoLabel: string | null
  subtotalCoins: number
  addonsCoins: number
  discountCoins: number
  totalCoins: number
}

export interface CancelPayload {
  reason: string
  refundOption: 'full' | 'partial' | 'none'
  refundCoins: number
  note: string
}

export interface DisputePayload {
  reason: string
  outcome: 'full' | 'partial' | 'reporting'
  note: string
}

export const useBookingsStore = defineStore('bookings', () => {
  const draft = ref<BookingDraft | null>(null)
  const current = ref<Booking | null>(null)

  const list = ref<Booking[]>([])
  const listLoading = ref(false)
  const listError = ref<string | null>(null)

  const incoming = ref<Booking[]>([])
  const incomingLoading = ref(false)
  const incomingError = ref<string | null>(null)

  function startDraft(payload: BookingDraft) {
    draft.value = payload
  }

  function replaceInPlace(booking: Booking) {
    for (const arr of [list.value, incoming.value]) {
      const index = arr.findIndex((b) => b.id === booking.id)
      if (index !== -1) arr[index] = booking
    }
    if (current.value?.id === booking.id) current.value = booking
  }

  function getBooking(id: string) {
    return (
      list.value.find((b) => b.id === id) ??
      incoming.value.find((b) => b.id === id) ??
      (current.value?.id === id ? current.value : null)
    )
  }

  /** My Bookings (buyer side). Falls back to `mockBookings` if the request fails (signed out,
   * network error, backend down) rather than showing a broken page. */
  async function fetchList() {
    listLoading.value = true
    listError.value = null
    try {
      list.value = await api.get<Booking[]>('/bookings/mine')
    } catch (err) {
      listError.value = err instanceof Error ? err.message : 'Failed to load orders'
      list.value = [...mockBookings]
    } finally {
      listLoading.value = false
    }
  }

  /** Pal Dashboard / Orders (Pal side). Same fallback behavior as `fetchList`. */
  async function fetchIncoming() {
    incomingLoading.value = true
    incomingError.value = null
    try {
      incoming.value = await api.get<Booking[]>('/bookings/incoming')
    } catch (err) {
      incomingError.value = err instanceof Error ? err.message : 'Failed to load orders'
      incoming.value = [...mockIncomingBookings]
    } finally {
      incomingLoading.value = false
    }
  }

  /** Order Detail direct/deep-link fallback when the booking isn't already in `list`/`incoming`
   * (e.g. a page refresh before either list was fetched). */
  async function fetchBooking(id: string) {
    current.value = await api.get<Booking>(`/bookings/${id}`)
    return current.value
  }

  /** Checkout's "Place order" - the point where a draft actually becomes a submitted booking. */
  async function placeOrder(payload: { paymentMethod: PaymentMethod; scheduledFor: string | null }) {
    if (!draft.value) throw new Error('No booking draft to submit')
    const booking = await api.post<Booking>('/bookings', { ...draft.value, ...payload })
    list.value.unshift(booking)
    current.value = booking
    draft.value = null
    return booking
  }

  async function acceptBooking(id: string) {
    const booking = await api.post<Booking>(`/bookings/${id}/accept`)
    replaceInPlace(booking)
    return booking
  }

  async function declineBooking(id: string) {
    const booking = await api.post<Booking>(`/bookings/${id}/decline`)
    replaceInPlace(booking)
    return booking
  }

  async function completeBooking(id: string) {
    const booking = await api.post<Booking>(`/bookings/${id}/complete`)
    replaceInPlace(booking)
    return booking
  }

  async function cancelBooking(id: string, payload: CancelPayload) {
    const booking = await api.post<Booking>(`/bookings/${id}/cancel`, payload)
    replaceInPlace(booking)
    return booking
  }

  async function reportIssue(id: string, payload: DisputePayload) {
    await api.post(`/bookings/${id}/dispute`, payload)
  }

  /** My Bookings' `LeaveReviewModal` submit (`POST /reviews`). `highlights`/`tipCoins` the modal
   * collects stay UI-only for now - see `routers/reviews.py`'s docstring. Patches `hasReview`
   * onto the local booking rather than refetching, same as the other action methods. */
  async function submitReview(bookingId: string, payload: ReviewPayload) {
    await api.post(`/reviews`, { bookingId, ...payload })
    const booking = getBooking(bookingId)
    if (booking) replaceInPlace({ ...booking, hasReview: true })
  }

  return {
    draft,
    current,
    list,
    listLoading,
    listError,
    incoming,
    incomingLoading,
    incomingError,
    startDraft,
    getBooking,
    fetchList,
    fetchIncoming,
    fetchBooking,
    placeOrder,
    acceptBooking,
    declineBooking,
    completeBooking,
    cancelBooking,
    reportIssue,
    submitReview,
  }
})

File: backend/src/backend/routers/bookings.py
import random
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import Field

from ..core.auth import get_current_user_id
from ..core.notify import notify
from ..core.schema import CamelModel
from ..core.supabase import get_supabase_client

router = APIRouter(prefix="/bookings", tags=["bookings"])


# Schemas ---------------------------------------------------------------------------------


class BookingAddonIn(CamelModel):
    label: str
    price_coins: int


class BookingAddonOut(CamelModel):
    id: str
    label: str
    price_coins: int


class BookingCreateIn(CamelModel):
    player_id: str
    service_id: str
    service_type_label: str
    price_coins: int
    price_unit: str
    quantity: int = 1
    addons: list[BookingAddonIn] = Field(default_factory=list)
    promo_label: str | None = None
    subtotal_coins: int
    addons_coins: int = 0
    discount_coins: int = 0
    total_coins: int
    payment_method: str = "coins"
    scheduled_for: str | None = None


class CancelIn(CamelModel):
    reason: str
    refund_option: str
    refund_coins: int = 0
    note: str | None = None


class DisputeIn(CamelModel):
    reason: str
    outcome: str
    note: str | None = None


class BookingOut(CamelModel):
    id: str
    order_number: str
    player_id: str
    player_display_name: str
    player_avatar_url: str | None
    service_id: str
    service_name: str
    service_type_label: str
    user_id: str
    buyer_display_name: str
    status: str
    price_coins: int
    price_unit: str
    quantity: int
    addons: list[BookingAddonOut]
    promo_label: str | None
    subtotal_coins: int
    addons_coins: int
    discount_coins: int
    total_coins: int
    payment_method: str
    scheduled_for: str | None
    created_at: str
    has_review: bool


# Helpers -----------------------------------------------------------------------------------

_SELECT = (
    "*, players(display_name, avatar_url, user_id), services(name), users(display_name), "
    "booking_addons(id, label, price_coins), reviews(id)"
)

_OUTCOME_MAP = {"full": "full_refund", "partial": "partial_refund", "reporting": "reporting"}


def _booking_out(row: dict) -> dict:
    player = row.get("players") or {}
    service = row.get("services") or {}
    buyer = row.get("users") or {}
    return {
        **row,
        "player_display_name": player.get("display_name") or "Pal",
        "player_avatar_url": player.get("avatar_url"),
        "service_name": service.get("name") or row["service_type_label"],
        "buyer_display_name": buyer.get("display_name") or "Buyer",
        "addons": row.get("booking_addons") or [],
        "has_review": bool(row.get("reviews")),
    }


def _booking_names(row: dict) -> tuple[str, str, str | None, str]:
    """(pal_name, buyer_name, pal_user_id, service_name) for building notification copy. Seed
    Pals have no `players.user_id` (2.3 note), so `pal_user_id` is `None` when there's nobody
    real to notify."""
    player = row.get("players") or {}
    service = row.get("services") or {}
    buyer = row.get("users") or {}
    pal_name = player.get("display_name") or "Your Pal"
    buyer_name = buyer.get("display_name") or "A buyer"
    service_name = service.get("name") or row["service_type_label"]
    return pal_name, buyer_name, player.get("user_id"), service_name


def _fetch_booking(client, booking_id: str) -> dict:
    result = client.table("bookings").select(_SELECT).eq("id", booking_id).maybe_single().execute()
    if not result or not result.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Order not found")
    return result.data


def _get_my_player_id(client, user_id: str) -> str | None:
    result = client.table("players").select("id").eq("user_id", user_id).maybe_single().execute()
    return result.data["id"] if result and result.data else None


def _get_owned_booking(client, booking_id: str, user_id: str) -> dict:
    """A booking reachable by whoever is party to it: the buyer (`user_id`) or the Pal
    fulfilling it (`players.user_id`)."""
    booking = _fetch_booking(client, booking_id)
    player_id = _get_my_player_id(client, user_id)
    if booking["user_id"] != user_id and booking["player_id"] != player_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Order not found")
    return booking


def _require_pal_booking(client, booking_id: str, user_id: str, *, expected_status: str) -> dict:
    player_id = _get_my_player_id(client, user_id)
    booking = _fetch_booking(client, booking_id)
    if not player_id or booking["player_id"] != player_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Order not found")
    if booking["status"] != expected_status:
        raise HTTPException(status.HTTP_409_CONFLICT, f"Order is not {expected_status}")
    return booking


def _generate_order_number(client) -> str:
    for _ in range(5):
        candidate = f"SQ-{random.randint(10000, 99999)}"
        existing = client.table("bookings").select("id").eq("order_number", candidate).maybe_single().execute()
        if not existing or not existing.data:
            return candidate
    raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, "Could not generate order number")


def _resolve_addon_id(client, label: str, price_coins: int) -> str:
    """Add-ons are still a flat, unauthored catalog (`mocks/bookings.ts`'s `mockAddons` on the
    frontend - see CHECKPOINT.md's Booking flow note), so there's no admin flow to pre-seed
    `addons` rows. Look one up by label, self-healing by inserting it the first time it's seen."""
    existing = client.table("addons").select("id").eq("label", label).maybe_single().execute()
    if existing and existing.data:
        return existing.data["id"]
    created = client.table("addons").insert({"label": label, "price_coins": price_coins}).execute()
    return created.data[0]["id"]


# Create + read -------------------------------------------------------------------------------


@router.post("", response_model=BookingOut, status_code=status.HTTP_201_CREATED)
def create_booking(payload: BookingCreateIn, user_id: str = Depends(get_current_user_id)) -> dict:
    """The buyer's "Place order" submission (Checkout). Note this is only called once Checkout
    actually submits, not when the "Book a session" modal's "Continue to checkout" fires - that
    stays a client-side draft (`stores/bookings.ts`'s `draft`) so an abandoned checkout never
    creates a real `pending` row, resolving the draft-vs-submitted question left open in
    CHECKPOINT.md's Booking flow note."""
    client = get_supabase_client()

    service = client.table("services").select("id, player_id").eq("id", payload.service_id).maybe_single().execute()
    if not service or not service.data or service.data["player_id"] != payload.player_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Service not found")

    booking_id = str(uuid4())
    order_number = _generate_order_number(client)

    client.table("bookings").insert(
        {
            "id": booking_id,
            "order_number": order_number,
            "player_id": payload.player_id,
            "service_id": payload.service_id,
            "user_id": user_id,
            "status": "pending",
            "service_type_label": payload.service_type_label,
            "price_coins": payload.price_coins,
            "price_unit": payload.price_unit,
            "quantity": payload.quantity,
            "promo_label": payload.promo_label,
            "subtotal_coins": payload.subtotal_coins,
            "addons_coins": payload.addons_coins,
            "discount_coins": payload.discount_coins,
            "total_coins": payload.total_coins,
            "payment_method": payload.payment_method,
            "scheduled_for": payload.scheduled_for,
        }
    ).execute()

    for addon in payload.addons:
        addon_id = _resolve_addon_id(client, addon.label, addon.price_coins)
        client.table("booking_addons").insert(
            {
                "booking_id": booking_id,
                "addon_id": addon_id,
                "label": addon.label,
                "price_coins": addon.price_coins,
            }
        ).execute()

    created = _fetch_booking(client, booking_id)
    _, buyer_name, pal_user_id, service_name = _booking_names(created)
    if pal_user_id:
        notify(pal_user_id, "booking", f"{buyer_name} requested to book {service_name}.")
    return _booking_out(created)


@router.get("/mine", response_model=list[BookingOut])
def list_my_bookings(user_id: str = Depends(get_current_user_id)) -> list[dict]:
    """My Bookings (buyer side)."""
    client = get_supabase_client()
    rows = (
        client.table("bookings")
        .select(_SELECT)
        .eq("user_id", user_id)
        .order("created_at", desc=True)
        .execute()
        .data
        or []
    )
    return [_booking_out(row) for row in rows]


@router.get("/incoming", response_model=list[BookingOut])
def list_incoming_bookings(user_id: str = Depends(get_current_user_id)) -> list[dict]:
    """Pal Dashboard / Orders (Pal side). Empty list rather than a 404 for an account with no
    Pal profile yet, since a plain user simply has no incoming orders."""
    client = get_supabase_client()
    player_id = _get_my_player_id(client, user_id)
    if not player_id:
        return []
    rows = (
        client.table("bookings")
        .select(_SELECT)
        .eq("player_id", player_id)
        .order("created_at", desc=True)
        .execute()
        .data
        or []
    )
    return [_booking_out(row) for row in rows]


@router.get("/{booking_id}", response_model=BookingOut)
def get_booking(booking_id: str, user_id: str = Depends(get_current_user_id)) -> dict:
    client = get_supabase_client()
    return _booking_out(_get_owned_booking(client, booking_id, user_id))


# Status transitions ----------------------------------------------------------------------


@router.post("/{booking_id}/accept", response_model=BookingOut)
def accept_booking(booking_id: str, user_id: str = Depends(get_current_user_id)) -> dict:
    client = get_supabase_client()
    _require_pal_booking(client, booking_id, user_id, expected_status="pending")
    client.table("bookings").update({"status": "accepted"}).eq("id", booking_id).execute()
    updated = _fetch_booking(client, booking_id)
    pal_name, _, _, service_name = _booking_names(updated)
    notify(updated["user_id"], "booking", f"{pal_name} accepted your booking for {service_name}.")
    return _booking_out(updated)


@router.post("/{booking_id}/decline", response_model=BookingOut)
def decline_booking(booking_id: str, user_id: str = Depends(get_current_user_id)) -> dict:
    client = get_supabase_client()
    _require_pal_booking(client, booking_id, user_id, expected_status="pending")
    client.table("bookings").update({"status": "declined"}).eq("id", booking_id).execute()
    updated = _fetch_booking(client, booking_id)
    pal_name, _, _, service_name = _booking_names(updated)
    notify(updated["user_id"], "booking", f"{pal_name} declined your booking for {service_name}.")
    return _booking_out(updated)


@router.post("/{booking_id}/complete", response_model=BookingOut)
def complete_booking(booking_id: str, user_id: str = Depends(get_current_user_id)) -> dict:
    client = get_supabase_client()
    _require_pal_booking(client, booking_id, user_id, expected_status="accepted")
    client.table("bookings").update({"status": "completed"}).eq("id", booking_id).execute()
    updated = _fetch_booking(client, booking_id)
    pal_name, _, _, service_name = _booking_names(updated)
    notify(updated["user_id"], "booking", f"Your session for {service_name} with {pal_name} is complete.")
    return _booking_out(updated)


@router.post("/{booking_id}/cancel", response_model=BookingOut)
def cancel_booking(booking_id: str, payload: CancelIn, user_id: str = Depends(get_current_user_id)) -> dict:
    """Reachable by either party (`CancelOrderModal` is currently only wired up on the buyer's
    My Bookings / Order Detail pages, but the reason list reads Pal-authored - see
    `components/modals/CancelOrderModal.vue` - so this doesn't assume which side is cancelling)."""
    client = get_supabase_client()
    booking = _get_owned_booking(client, booking_id, user_id)
    if booking["status"] not in ("pending", "accepted"):
        raise HTTPException(status.HTTP_409_CONFLICT, "Order can no longer be cancelled")

    client.table("order_cancellations").insert(
        {
            "booking_id": booking_id,
            "cancelled_by": user_id,
            "reason": payload.reason,
            "refund_option": payload.refund_option,
            "refund_coins": payload.refund_coins,
            "note": payload.note,
        }
    ).execute()
    client.table("bookings").update({"status": "declined"}).eq("id", booking_id).execute()
    updated = _fetch_booking(client, booking_id)
    pal_name, buyer_name, pal_user_id, service_name = _booking_names(updated)
    if user_id == updated["user_id"]:
        # the buyer cancelled - notify the Pal
        if pal_user_id:
            notify(pal_user_id, "booking", f"{buyer_name} cancelled the booking for {service_name}.")
    else:
        # the Pal cancelled - notify the buyer
        notify(updated["user_id"], "booking", f"{pal_name} cancelled the booking for {service_name}.")
    return _booking_out(updated)


@router.post("/{booking_id}/dispute", status_code=status.HTTP_201_CREATED)
def dispute_booking(booking_id: str, payload: DisputeIn, user_id: str = Depends(get_current_user_id)) -> dict:
    """Buyer's "Report an issue" (`RefundModal`). Backs both that flow and the admin Disputes tab
    - see CHECKPOINT.md's 2.3 note on `order_disputes`."""
    client = get_supabase_client()
    booking = _fetch_booking(client, booking_id)
    if booking["user_id"] != user_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Order not found")

    requested_outcome = _OUTCOME_MAP.get(payload.outcome, "reporting")
    client.table("order_disputes").insert(
        {
            "booking_id": booking_id,
            "reported_by": user_id,
            "reason": payload.reason,
            "requested_outcome": requested_outcome,
            "refund_coins": booking["total_coins"] if requested_outcome == "full_refund" else None,
            "note": payload.note,
        }
    ).execute()
    return {"status": "submitted"}
